src/lib/helpers.ts

```diff
@@ -282,3 +282,112 @@ export async function generatePKCEChallenge(verifier: string) {
   const hashed = await sha256(verifier)
   return base64urlencode(hashed)
 }
+
+const STACK_GUARD_PREFIX = `__stack_guard__`
+const STACK_GUARD_SUFFIX = `__`
+
+// Firefox and WebKit based browsers encode the stack entry differently, but
+// they all include the function name. So instead of trying to parse the entry,
+// we're only looking for the special string `__stack_guard__${guardName}`. The
+// guard name is terminated when seeing '@', ' ' or '('.
+//
+// Example Firefox stack trace:
+// ```
+// __stack_guard__EXAMPLE@debugger eval code:1:55
+// @debugger eval code:1:3
+// ```
+//
+// Example WebKit/Chrome stack trace:
+// ```
+// Error
+//  at Object.__stack_guard__EXAMPLE (<anonymous>:1:55)
+//  at <anonymous>:1:13
+// ```
+//
+const STACK_ENTRY_REGEX = /__stack_guard__([a-zA-Z_-]+)__/
+
+let STACK_GUARD_CHECKED = false
+let STACK_GUARD_CHECK_FN: () => Promise<void> // eslint-disable-line prefer-const
+
+/**
+ * Checks if the current caller of the function is in a {@link
+ * #stackGuard} of the provided `name`. Works by looking through
+ * the stack trace of an `Error` object for a special function
+ * name (generated by {@link #stackGuard}).
+ *
+ * @param name The name of the stack guard to check for. Must be `[a-zA-Z_-]` only.
+ */
+export function isInStackGuard(name: string): boolean {
+  STACK_GUARD_CHECK_FN()
```

Comment:
hmm what is the point of this?

Reply:
To warn users that the environment they're running the library in does not support stack guards due to either transpilation to JS that uses generators (<= ES2015) or in a JavaScript engine that does not support async/await stack traces. 

This will very rarely be encountered, as only some fringe JS engines don't support async/await stack traces.

Reply:
I updated the warning message to be more instructive.